### PR TITLE
Add tooling to explain why a graph execution happens

### PR DESCRIPTION
**Pull request by JackCaoG**
FYI @Liyang90 @AlexWertheim

example output
```
Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   mark_step in parallel loader at step end
Execution Analysis: Python Frame Triggered Execution: 
Execution Analysis:   mark_step (/src/pytorch/xla/torch_xla/core/xla_model.py:818)
Execution Analysis:   next (/src/pytorch/xla/torch_xla/distributed/parallel_loader.py:44)
Execution Analysis:   __next__ (/src/pytorch/xla/torch_xla/distributed/parallel_loader.py:32)
Execution Analysis:   train_loop_fn (/src/pytorch/xla/test/test_train_mp_imagenet.py:296)
Execution Analysis:   train_imagenet (/src/pytorch/xla/test/test_train_mp_imagenet.py:345)
Execution Analysis:   _mp_fn (/src/pytorch/xla/test/test_train_mp_imagenet.py:369)
Execution Analysis:   __call__ (/src/pytorch/xla/torch_xla/_internal/pjrt.py:189)
Execution Analysis:   _thread_fn (/src/pytorch/xla/torch_xla/_internal/pjrt.py:70)
Execution Analysis:   run (/usr/local/lib/python3.8/concurrent/futures/thread.py:57)
Execution Analysis:   _worker (/usr/local/lib/python3.8/concurrent/futures/thread.py:80)
Execution Analysis:   run (/usr/local/lib/python3.8/threading.py:870)
Execution Analysis:   _bootstrap_inner (/usr/local/lib/python3.8/threading.py:932)
Execution Analysis:   _bootstrap (/usr/local/lib/python3.8/threading.py:890)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================
Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   mark_step when existing a profiler StepTrace region
Execution Analysis: Python Frame Triggered Execution: 
Execution Analysis:   mark_step (/src/pytorch/xla/torch_xla/core/xla_model.py:818)
Execution Analysis:   __exit__ (/src/pytorch/xla/torch_xla/debug/profiler.py:160)
Execution Analysis:   train_loop_fn (/src/pytorch/xla/test/test_train_mp_imagenet.py:311)
Execution Analysis:   train_imagenet (/src/pytorch/xla/test/test_train_mp_imagenet.py:345)
Execution Analysis:   _mp_fn (/src/pytorch/xla/test/test_train_mp_imagenet.py:369)
Execution Analysis:   __call__ (/src/pytorch/xla/torch_xla/_internal/pjrt.py:189)
Execution Analysis:   _thread_fn (/src/pytorch/xla/torch_xla/_internal/pjrt.py:70)
Execution Analysis:   run (/usr/local/lib/python3.8/concurrent/futures/thread.py:57)
Execution Analysis:   _worker (/usr/local/lib/python3.8/concurrent/futures/thread.py:80)
Execution Analysis:   run (/usr/local/lib/python3.8/threading.py:870)
Execution Analysis:   _bootstrap_inner (/usr/local/lib/python3.8/threading.py:932)
Execution Analysis:   _bootstrap (/usr/local/lib/python3.8/threading.py:890)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================
```

### CODE REVIEW

File: torch_xla/csrc/debug_util.cpp [hunk 1]
```diff
@@ -209,4 +212,43 @@ bool DebugUtil::ExperimentEnabled(const std::string& name) {
...
+  std::vector<torch::lazy::SourceLocation> frames =
+      torch::lazy::GetPythonFrames();
+  // python frame must be > 1
+  XLA_CHECK_GE(frames.size(), 1);
+  std::stringstream ss;
+  ss << debug_output_prefix
+     << "======================================================================"
+        "=========="
+     << "\n";
+  if (frames[0].function == "mark_step") {
```
Comment (vanbasten23) [inline torch_xla/csrc/debug_util.cpp:227]: Wondering if it worths logging `frames[0].function;` for the cases other than `mark_step`.

**Review by vanbasten23 — APPROVED** [review]

**Comment by JackCaoG** [comment]
I need to add some test, will try to do that this week

File: torch_xla/csrc/debug_util.cpp (hunk 1 above)
Comment (JackCaoG) [reply torch_xla/csrc/debug_util.cpp:239]: yea I want to expand this a bit later to cover most of the common cases of the execution.

**Comment by JackCaoG** [comment]
This should be ready for review.

**Comment by JackCaoG** [comment]
I will update troubleshotting in a separate pr.

**Comment by JackCaoG** [comment]
@will-cromar Can I get a review for this one?

**Review by alanwaketan — COMMENTED** [review]
I think this information is already in the dumped graphs?

Will the UX be better with this change where the information is printed to the console? It may not be the case as the information will be interleaved with the user code.

**Comment by JackCaoG** [comment]
dumped graph is in a separate file and it always dump the HLO or IR which can be huge. The idea is that we should dump something more concise and provide some explanation on what is happening, this can be interleave with the debugging message user added to their model code.

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Comment by JackCaoG** [comment]
@carmocca FYI, this is a WIP I will add more to this tooling.

**Pull request merged**